File: backend/cloudflare-api/scoring.py
"""Reiknar norðurljósaskor (0-100) fyrir einn stað út frá gögnum sem
sources.py hefur sótt. Vægi fyrir í kvöld (staðbundin OVATION-mæling til):
virkni 35%, Kp-leitni 10%, heiðskírt 40%, tunglbirta 15%. Fyrir síðari
nætur er engin staðbundin virknimæling til (OVATION er bara núgildisspá)
- þá fellur virknivægið (45% samtals) alfarið á Kp-spána, sem er ein tala
fyrir allt landið, ekki staðbundin. Skýjahula vegur þyngst því norðurljós
sjást einfaldlega ekki gegnum ský, óháð virkni."""

from datetime import datetime, timedelta, timezone


def ovation_virkni(lat, lon, ovation_data):
    if not ovation_data:
        return None
    lon_0_360 = lon % 360
    lat_r = round(lat)
    lon_r = round(lon_0_360)
    besta = None
    for lo, la, v in ovation_data["coordinates"]:
        if la == lat_r and lo == lon_r:
            return v
        d = abs(la - lat) + min(abs(lo - lon_0_360), 360 - abs(lo - lon_0_360))
        if besta is None or d < besta[0]:
            besta = (d, v)
    return besta[1] if besta else None


def kp_leitni(kp_spa, myrkur_fra, myrkur_til):
    if not kp_spa or not myrkur_fra or not myrkur_til:
        return None
    gildi = []
    for row in kp_spa:
        if row.get("observed") not in ("predicted", "estimated"):
            continue
        try:
            t = datetime.fromisoformat(row["time_tag"]).replace(tzinfo=timezone.utc)
        except ValueError:
            continue
        if myrkur_fra <= t <= myrkur_til:
            gildi.append(float(row["kp"]))
    return sum(gildi) / len(gildi) if gildi else None


def medal_skyjahula(skyjahula, myrkur_fra, myrkur_til):
    """Meðal heildarskýjahula (%) innan myrkurgluggans - notað til
    BIRTINGAR ('Skýjahula 62%'), ekki í skorreikninginn sjálfan."""
    if not skyjahula or not myrkur_fra or not myrkur_til:
        return None
    gildi = [
        p["heild"] for t, p in skyjahula.items()
        if myrkur_fra <= t <= myrkur_til and p.get("heild") is not None
    ]
    return sum(gildi) / len(gildi) if gildi else None


def medal_skyjaopacitet(skyjahula, myrkur_fra, myrkur_til):
    """Vegið meðaltal skýjaþéttleika innan myrkurgluggans, notað í sjálfan
    skorreikninginn. Lágský vega þyngst (60%), miðlungsský næst (30%) og
    háský minnst (10%) - þunn háský hindra norðurljós miklu minna en þétt
    lágský, svo flatt heildarhulu-meðaltal ofmetur áhrif háskýja."""
    if not skyjahula or not myrkur_fra or not myrkur_til:
        return None
    gildi = []
    for t, p in skyjahula.items():
        if not (myrkur_fra <= t <= myrkur_til):
            continue
        lagt, midlungs, hatt = p.get("lagt"), p.get("midlungs"), p.get("hatt")
        if lagt is None and midlungs is None and hatt is None:
            if p.get("heild") is not None:
                gildi.append(p["heild"])
            continue
        opacitet = 0.60 * (lagt or 0) + 0.30 * (midlungs or 0) + 0.10 * (hatt or 0)
        gildi.append(min(opacitet, 100))
    return sum(gildi) / len(gildi) if gildi else None


def haestu_uv_dagsins(skyjahula, dags):
    """Hæsta UV-gildi dagsins (ekki myrkurgluggans), leiðrétt gróflega
    fyrir raunverulega skýjahulu (UV-gildið frá MET er miðað við heiðskírt)."""
    if not skyjahula or not dags:
        return None
    upphaf = datetime.combine(dags, datetime.min.time(), tzinfo=timezone.utc)
    lok = upphaf + timedelta(days=1)
    gildi = []
    for t, p in skyjahula.items():
        if not (upphaf <= t < lok) or p.get("uv_heidskirt") is None:
            continue
        heild = p.get("heild")
        leidrett = p["uv_heidskirt"] * (1 - (heild if heild is not None else 50) / 100)
        gildi.append(leidrett)
    return round(max(gildi), 1) if gildi else None


def _tungl_bil_fyrir_dag(dagsfaersla):
    """Skilar lista af (upphaf, endir) tímabilum sem tunglið er yfir
    sjóndeildarhring innan almanaksdagsins sem dagsfaersla lýsir. Höndlar
    daga þar sem tunglið sest fyrir uppkomu (uppi um miðnætti) og daga þar
    sem annaðhvort uppkoma eða setur fellur ekki á daginn sjálfan."""
    dags_upphaf = datetime.combine(dagsfaersla["dags"], datetime.min.time(), tzinfo=timezone.utc)
    dags_lok = dags_upphaf + timedelta(days=1)
    r = dagsfaersla["tungl_upp"]
    s = dagsfaersla["tungl_nidur"]

    if dagsfaersla["tungl_alltaf_uppi"]:
        return [(dags_upphaf, dags_lok)]
    if dagsfaersla["tungl_alltaf_nidri"]:
        return []
    if r and s:
        return [(r, s)] if r <= s else [(dags_upphaf, s), (r, dags_lok)]
    if r and not s:
        return [(r, dags_lok)]
    if s and not r:
        return [(dags_upphaf, s)]
    return []


def tungl_uppi_hlutfall(dagar_gogn, dagur, myrkur_fra, myrkur_til):
    """Hlutfall (0-1) af myrkurglugganum sem tunglið er yfir sjóndeildar-
    hring. Notað til að refsa aðeins fyrir tunglbirtu þann tíma sem
    tunglið er raunverulega uppi - ekki allan gluggann óháð því."""
    if not dagar_gogn or not myrkur_fra or not myrkur_til or myrkur_til <= myrkur_fra:
        return None

    bil = []
    for i in (dagur, dagur + 1):
        if i < len(dagar_gogn):
            bil.extend(_tungl_bil_fyrir_dag(dagar_gogn[i]))

    heild_sek = 0.0
    for upphaf, endir in bil:
        skorun_upphaf = max(upphaf, myrkur_fra)
        skorun_endir = min(endir, myrkur_til)
        if skorun_endir > skorun_upphaf:
            heild_sek += (skorun_endir - skorun_upphaf).total_seconds()

    lengd_sek = (myrkur_til - myrkur_fra).total_seconds()
    return min(1.0, heild_sek / lengd_sek) if lengd_sek > 0 else None


def hiti_vid_myrkur(skyjahula, myrkur_fra):
    """Hitastig og 'finnst eins og' á þeim tímapunkti sem næst kemur upphafi
    myrkurgluggans (dusk) - eingöngu til upplýsinga, hefur ekki áhrif á
    skorið sjálft."""
    if not skyjahula or not myrkur_fra:
        return None, None
    naestur = min(skyjahula.items(), key=lambda kv: abs((kv[0] - myrkur_fra).total_seconds()))
    p = naestur[1]
    return p.get("hiti"), p.get("hiti_finnst")


def naestigildi(skyjahula, timi):
    """Næsta klukkustundargildi (heild/hiti/vindur o.fl.) við tiltekinn
    tímapunkt - almennt uppflettifall fyrir vaktarskjáinn."""
    if not skyjahula or not timi:
        return {}
    naestur = min(skyjahula.items(), key=lambda kv: abs((kv[0] - timi).total_seconds()))
    return naestur[1]


def naeturklukkustundir(fra, til):
    """Listi af heilum klukkustundum (dagsetningar-hlutir) frá og með
    fyrstu heilu stund eftir 'fra' til og með síðustu heilu stund fyrir
    eða á 'til'."""
    if not fra or not til or til <= fra:
        return []
    fyrsta = fra.replace(minute=0, second=0, microsecond=0)
    if fyrsta < fra:
        fyrsta += timedelta(hours=1)
    ut = []
    t = fyrsta
    while t <= til:
        ut.append(t)
        t += timedelta(hours=1)
    return ut


def kp_a_klukkustund(kp_spa, klst_timar):
    """Kp-spá NOAA er á 3ja klst upplausn - línuleg brúun gefur eina tölu
    fyrir hverja klukkustund í klst_timar."""
    if not klst_timar:
        return []
    if not kp_spa:
        return [None] * len(klst_timar)

    punktar = []
    for row in kp_spa:
        try:
            t = datetime.fromisoformat(row["time_tag"]).replace(tzinfo=timezone.utc)
            punktar.append((t, float(row["kp"])))
        except (ValueError, KeyError, TypeError):
            continue
    punktar.sort(key=lambda p: p[0])
    if not punktar:
        return [None] * len(klst_timar)

    ut = []
    for klst in klst_timar:
        if klst <= punktar[0][0]:
            ut.append(round(punktar[0][1], 2))
            continue
        if klst >= punktar[-1][0]:
            ut.append(round(punktar[-1][1], 2))
            continue
        for i in range(len(punktar) - 1):
            t0, v0 = punktar[i]
            t1, v1 = punktar[i + 1]
            if t0 <= klst <= t1:
                hlutfall = (klst - t0).total_seconds() / (t1 - t0).total_seconds()
                ut.append(round(v0 + (v1 - v0) * hlutfall, 2))
                break
    return ut


def sky_a_klukkustund(skyjahula, klst_timar):
    """Heildarskýjahula (%) fyrir hverja klukkustund - notað í súlurit."""
    return [
        (round(p["heild"]) if p.get("heild") is not None else None)
        for p in (naestigildi(skyjahula, klst) for klst in klst_timar)
    ]


TUNGLFASA_HEITI = {
    "New Moon": "Nýtt tungl",
    "Waxing Crescent": "Vaxandi mánasigð",
    "First Quarter": "Fyrsta kvartil",
    "Waxing Gibbous": "Vaxandi tungl",
    "Full Moon": "Fullt tungl",
    "Waning Gibbous": "Dvínandi tungl",
    "Last Quarter": "Síðasta kvartil",
    "Waning Crescent": "Dvínandi mánasigð",
}


def solvindur_tulkun(bz, hradi):
    """Einföld, mannlesanleg túlkun á núverandi sólvindsaðstæðum - eingöngu
    til upplýsinga (OVATION er þegar reiknað út frá þessum sömu gögnum,
    svo þetta breytir ekki skorinu)."""
    if bz is None:
        return "Gögn um sólvind ekki tiltæk í augnablikinu."

    if bz <= -10:
        domur = "mjög hagstætt fyrir norðurljós núna"
    elif bz <= -5:
        domur = "hagstætt fyrir norðurljós núna"
    elif bz < 0:
        domur = "hlutlaust til hagstætt núna"
    elif bz < 5:
        domur = "hlutlaust núna"
    else:
        domur = "óhagstætt fyrir norðurljós núna"

    setning = f"Bz-gildið er {bz:.1f} nT — {domur}."
    if hradi is not None:
        if hradi >= 500:
            setning += " Sólvindshraði er hár, sem magnar áhrifin."
        elif hradi < 350:
            setning += " Sólvindshraði er lágur."
    return setning


def reikna_skor(virkni, kp, sky_opacitet, tungl_pct, tungl_uppi, myrkur_fra, myrkur_til):
    if myrkur_fra is None or myrkur_til is None or myrkur_til <= myrkur_fra:
        return {"skor": 0.0, "ástæða": "ekkert marktækt myrkur á tímabilinu"}

    heidskirt_stig = 100 - (sky_opacitet if sky_opacitet is not None else 50)
    tungl_birta = tungl_pct if tungl_pct is not None else 30
    tungl_upp_hlutf = tungl_uppi if tungl_uppi is not None else 1.0
    tungl_stig = 100 - (tungl_birta * tungl_upp_hlutf)
    kp_stig = min((kp or 0) / 9.0, 1.0) * 100 if kp is not None else 50.0

    if virkni is not None:
        # virkni er nú þegar prósentulíkur (0-100) á sýnilegum norðurljósum
        # skv. NOAA - ekki tala á 0-14 kvarða eins og upprunalega skráin gerði ráð fyrir.
        virkni_stig = min(max(virkni, 0.0), 100.0)
        skor = 0.35 * virkni_stig + 0.10 * kp_stig + 0.40 * heidskirt_stig + 0.15 * tungl_stig
        nakvaemni = "mæling"
    else:
        virkni_stig = None
        skor = 0.45 * kp_stig + 0.40 * heidskirt_stig + 0.15 * tungl_stig
        nakvaemni = "spá"

    return {
        "skor": round(max(0.0, min(100.0, skor)), 1),
        "nakvaemni": nakvaemni,
        "sundurlidun": {
            "virkni_stig": round(virkni_stig, 1) if virkni_stig is not None else None,
            "kp_stig": round(kp_stig, 1),
            "heidskirt_stig": round(heidskirt_stig, 1),
            "tungl_stig": round(tungl_stig, 1),
        },
    }

File: cloudflare-api/src/scoring.js
// Reiknar norðurljósaskor (0-100) fyrir einn stað. Bein þýðing á scoring.py.
// Tímar eru millisekúndur (UTC); skýjagögn eru listi af [ms, gildi] í sömu röð og í KV.

import { HOUR, pyRound, naiveUtc, dateStartMs, mean } from "./pyutil.js";

export function ovationVirkni(lat, lon, ovationData) {
  if (!ovationData) return null;
  const lon360 = ((lon % 360) + 360) % 360;
  const latR = pyRound(lat);
  const lonR = pyRound(lon360);
  let besta = null;
  for (const [lo, la, v] of ovationData.coordinates) {
    if (la === latR && lo === lonR) return v;
    const d = Math.abs(la - lat) + Math.min(Math.abs(lo - lon360), 360 - Math.abs(lo - lon360));
    if (besta === null || d < besta[0]) besta = [d, v];
  }
  return besta ? besta[1] : null;
}

export function kpLeitni(kpSpa, myrkurFra, myrkurTil) {
  if (!kpSpa || !kpSpa.length || !myrkurFra || !myrkurTil) return null;
  const gildi = [];
  for (const row of kpSpa) {
    if (row.observed !== "predicted" && row.observed !== "estimated") continue;
    const t = naiveUtc(row.time_tag);
    if (t === null) continue;
    if (myrkurFra <= t && t <= myrkurTil) gildi.push(Number(row.kp));
  }
  return mean(gildi);
}

export function medalSkyjahula(sky, myrkurFra, myrkurTil) {
  if (!sky || !sky.length || !myrkurFra || !myrkurTil) return null;
  const gildi = [];
  for (const [t, p] of sky) {
    if (myrkurFra <= t && t <= myrkurTil && p.heild !== null && p.heild !== undefined) gildi.push(p.heild);
  }
  return mean(gildi);
}

export function medalSkyjaopacitet(sky, myrkurFra, myrkurTil) {
  if (!sky || !sky.length || !myrkurFra || !myrkurTil) return null;
  const gildi = [];
  for (const [t, p] of sky) {
    if (!(myrkurFra <= t && t <= myrkurTil)) continue;
    const lagt = p.lagt ?? null, midlungs = p.midlungs ?? null, hatt = p.hatt ?? null;
    if (lagt === null && midlungs === null && hatt === null) {
      if (p.heild !== null && p.heild !== undefined) gildi.push(p.heild);
      continue;
    }
    const opacitet = 0.6 * (lagt || 0) + 0.3 * (midlungs || 0) + 0.1 * (hatt || 0);
    gildi.push(Math.min(opacitet, 100));
  }
  return mean(gildi);
}

export function haestuUvDagsins(sky, dags) {
  if (!sky || !sky.length || !dags) return null;
  const upphaf = dateStartMs(dags);
  const lok = upphaf + 24 * HOUR;
  const gildi = [];
  for (const [t, p] of sky) {
    if (!(upphaf <= t && t < lok) || p.uv_heidskirt === null || p.uv_heidskirt === undefined) continue;
    const heild = p.heild ?? null;
    gildi.push(p.uv_heidskirt * (1 - (heild !== null ? heild : 50) / 100));
  }
  return gildi.length ? pyRound(Math.max(...gildi), 1) : null;
}

function tunglBilFyrirDag(d) {
  const upphaf = dateStartMs(d.dags);
  const lok = upphaf + 24 * HOUR;
  const r = d.tungl_upp;
  const s = d.tungl_nidur;
  if (d.tungl_alltaf_uppi) return [[upphaf, lok]];
  if (d.tungl_alltaf_nidri) return [];
  if (r && s) return r <= s ? [[r, s]] : [[upphaf, s], [r, lok]];
  if (r && !s) return [[r, lok]];
  if (s && !r) return [[upphaf, s]];
  return [];
}

export function tunglUppiHlutfall(dagar, dagur, myrkurFra, myrkurTil) {
  if (!dagar || !dagar.length || !myrkurFra || !myrkurTil || myrkurTil <= myrkurFra) return null;
  const bil = [];
  for (const i of [dagur, dagur + 1]) {
    if (i < dagar.length) bil.push(...tunglBilFyrirDag(dagar[i]));
  }
  let heild = 0;
  for (const [upphaf, endir] of bil) {
    const a = Math.max(upphaf, myrkurFra);
    const b = Math.min(endir, myrkurTil);
    if (b > a) heild += (b - a) / 1000;
  }
  const lengd = (myrkurTil - myrkurFra) / 1000;
  return lengd > 0 ? Math.min(1.0, heild / lengd) : null;
}

export function naestigildi(sky, timi) {
  if (!sky || !sky.length || !timi) return {};
  let best = sky[0];
  for (const item of sky) {
    if (Math.abs(item[0] - timi) < Math.abs(best[0] - timi)) best = item;
  }
  return best[1];
}

export function hitiVidMyrkur(sky, myrkurFra) {
  if (!sky || !sky.length || !myrkurFra) return [null, null];
  const p = naestigildi(sky, myrkurFra);
  return [p.hiti ?? null, p.hiti_finnst ?? null];
}

export function naeturklukkustundir(fra, til) {
  if (!fra || !til || til <= fra) return [];
  let t = Math.floor(fra / HOUR) * HOUR;
  if (t < fra) t += HOUR;
  const ut = [];
  for (; t <= til; t += HOUR) ut.push(t);
  return ut;
}

export function kpAKlukkustund(kpSpa, klstTimar) {
  if (!klstTimar.length) return [];
  if (!kpSpa || !kpSpa.length) return klstTimar.map(() => null);
  const punktar = [];
  for (const row of kpSpa) {
    if (!row || row.kp === null || row.kp === undefined) continue;
    const t = naiveUtc(row.time_tag);
    const v = Number(row.kp);
    if (t === null || Number.isNaN(v)) continue;
    punktar.push([t, v]);
  }
  punktar.sort((a, b) => a[0] - b[0]);
  if (!punktar.length) return klstTimar.map(() => null);
  const ut = [];
  for (const klst of klstTimar) {
    if (klst <= punktar[0][0]) { ut.push(pyRound(punktar[0][1], 2)); continue; }
    const last = punktar[punktar.length - 1];
    if (klst >= last[0]) { ut.push(pyRound(last[1], 2)); continue; }
    for (let i = 0; i < punktar.length - 1; i++) {
      const [t0, v0] = punktar[i];
      const [t1, v1] = punktar[i + 1];
      if (t0 <= klst && klst <= t1) {
        const hlutfall = (klst - t0) / (t1 - t0);
        ut.push(pyRound(v0 + (v1 - v0) * hlutfall, 2));
        break;
      }
    }
  }
  return ut;
}

export function skyAKlukkustund(sky, klstTimar) {
  return klstTimar.map((klst) => {
    const p = naestigildi(sky, klst);
    return p.heild !== null && p.heild !== undefined ? pyRound(p.heild) : null;
  });
}

export const TUNGLFASA_HEITI = {
  "New Moon": "Nýtt tungl",
  "Waxing Crescent": "Vaxandi mánasigð",
  "First Quarter": "Fyrsta kvartil",
  "Waxing Gibbous": "Vaxandi tungl",
  "Full Moon": "Fullt tungl",
  "Waning Gibbous": "Dvínandi tungl",
  "Last Quarter": "Síðasta kvartil",
  "Waning Crescent": "Dvínandi mánasigð",
};

export function solvindurTulkun(bz, hradi) {
  if (bz === null || bz === undefined) return "Gögn um sólvind ekki tiltæk í augnablikinu.";
  let domur;
  if (bz <= -10) domur = "mjög hagstætt fyrir norðurljós núna";
  else if (bz <= -5) domur = "hagstætt fyrir norðurljós núna";
  else if (bz < 0) domur = "hlutlaust til hagstætt núna";
  else if (bz < 5) domur = "hlutlaust núna";
  else domur = "óhagstætt fyrir norðurljós núna";

  let bzText = Math.abs(pyRound(bz, 1)).toFixed(1);
  if (bz < 0) bzText = "-" + bzText;
  let setning = `Bz-gildið er ${bzText} nT — ${domur}.`;
  if (hradi !== null && hradi !== undefined) {
    if (hradi >= 500) setning += " Sólvindshraði er hár, sem magnar áhrifin.";
    else if (hradi < 350) setning += " Sólvindshraði er lágur.";
  }
  return setning;
}

export function reiknaSkor(virkni, kp, skyOpacitet, tunglPct, tunglUppi, myrkurFra, myrkurTil) {
  if (myrkurFra === null || myrkurTil === null || myrkurTil <= myrkurFra) {
    return { skor: 0.0, "ástæða": "ekkert marktækt myrkur á tímabilinu" };
  }
  const heidskirtStig = 100 - (skyOpacitet !== null ? skyOpacitet : 50);
  const tunglBirta = tunglPct !== null ? tunglPct : 30;
  const tunglUppHlutf = tunglUppi !== null ? tunglUppi : 1.0;
  const tunglStig = 100 - tunglBirta * tunglUppHlutf;
  const kpStig = kp !== null ? Math.min((kp || 0) / 9.0, 1.0) * 100 : 50.0;

  let virkniStig, skor, nakvaemni;
  if (virkni !== null && virkni !== undefined) {
    // virkni er prósentulíkur (0-100) á sýnilegum norðurljósum skv. NOAA.
    virkniStig = Math.min(Math.max(virkni, 0.0), 100.0);
    skor = 0.35 * virkniStig + 0.1 * kpStig + 0.4 * heidskirtStig + 0.15 * tunglStig;
    nakvaemni = "mæling";
  } else {
    virkniStig = null;
    skor = 0.45 * kpStig + 0.4 * heidskirtStig + 0.15 * tunglStig;
    nakvaemni = "spá";
  }
  return {
    skor: pyRound(Math.max(0.0, Math.min(100.0, skor)), 1),
    nakvaemni,
    sundurlidun: {
      virkni_stig: virkniStig !== null ? pyRound(virkniStig, 1) : null,
      kp_stig: pyRound(kpStig, 1),
      heidskirt_stig: pyRound(heidskirtStig, 1),
      tungl_stig: pyRound(tunglStig, 1),
    },
  };
}
